[WIP] Drivers: CAN: Implementation of a send timeout for can_send(...)

alexanderwachter opened the pull request:
This PR implements a timeout for outgoing CAN frames.

~~At the moment only the stm32 driver implements the new functionality. Other drivers will follow by the end of the week.~~

It fixes issue #19502

Commit:
drivers: CAN: stm32: add send timeout

This commit implements the API change where a timeout for sending
frames is added. Send requests that exceed the timeout are aborted.
The callback is issued when the corresponding mailbox gets empty
again.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
drivers: CAN: Change u32_t error_flags to int error_numbers

The error_flags parameter of the can_tx_callback_t is a u32_t
but gets negative CAN_TX error codes. This commit changes the
error_flags to int error_numbers.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
ping @nixward

alexanderwachter commented on drivers/can/can_stm32.c:
```diff
@@ -413,16 +441,34 @@ int can_stm32_send(struct device *dev, const struct zcan_frame *msg,
+			    mb_timeout);
+		time_spent = k_uptime_get() - time_stamp;
+		if (time_spent >= mb_timeout || mb_timeout == K_NO_WAIT ||
+		    k_sem_take(&data->tx_int_sem, mb_timeout - time_spent)) {
```
I'm not sure if `mb_timeout - time_spent` is allowed as timeout parameter here since I assume milliseconds and not use K_MSEC(). Maybe @pabigot can clarify.

pabigot replied:
This is exactly the sort of code that can't be automatically converted without risk.  The question is whether the driver wants to maintain `mb_timeout` as a count of milliseconds, or as a value to be passed to the kernel API.

Looking at the way the driver seems to do things, and assuming there's no need for timeouts measured in microseconds or anything finer than a millisecond, I would say that the variables used to pass timeouts should remain integers, documented as being specified in milliseconds.

In this code that would mean that `mb_timeout == K_NO_WAIT` would become `mb_timeout == 0`, and the value passed to `k_sem_take` would be `K_MSEC(mb_timeout - time_spent)`.

karstenkoenig commented on drivers/can/can_stm32.c:
```diff
@@ -468,11 +514,24 @@ int can_stm32_send(struct device *dev, const struct zcan_frame *msg,
 	if (callback == NULL) {
-		k_sem_take(&mb->tx_int_sem, K_FOREVER);
-		return mb->error_flags;
+		k_sem_init(&mb->tx_int_sem, 0, 1);
```
Should you really initialize the semaphore here? What happens if the transmission already happened before this point in time and the tx interrupt has already fired?

alexanderwachter replied:
mb semaphores are removed now.

alexanderwachter committed:
drivers: CAN: Abstract blocking send call

This commit moves the semaphore for a blocking call from
the driver implementation to the stack of the caller.
It reduces the complexity of the drivers and saves some RAM.

Signed-off-by: Alexander Wachter <alexander@wachter.cloud>

erwango commented:
@alexanderwachter any plans for this PR ?

alexanderwachter commented:
@erwango the plan is to get it in somewhen. But my priority is on the CAN-FD support ath the moment.

pabigot commented:
Could be superseded by #24511 if global driver timeouts get traction.

alexanderwachter commented:
@pabigot I would love to see a common solution for all APIs. Never the less, this PR implements the mechanisms to cancel a transmission.
